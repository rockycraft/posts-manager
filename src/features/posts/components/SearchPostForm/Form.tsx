import { memo } from "react";
import { FormProps } from "@/types/FormProps";
import { Button, HStack, Input } from "@chakra-ui/react";
import { HiMagnifyingGlass } from "react-icons/hi2";

function Form({ onSubmit, register }: FormProps<{ name: string }>) {
	return (
		<form role="form" onSubmit={onSubmit}>
			<HStack>
				<Input
					data-testid="name-search-input"
					placeholder="Nombre"
					type="text"
					{...register("name")}
					variant="subtle"
				/>
				<Button data-testid="submit-button" type="submit" colorPalette="blue">
					<HiMagnifyingGlass />
					Buscar
				</Button>
			</HStack>
		</form>
	);
}

export default memo(Form);
