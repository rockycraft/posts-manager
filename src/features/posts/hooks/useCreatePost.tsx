import type { State } from "@/types/State";
import { useCallback, useState } from "react";
import { useDispatch } from "react-redux";
import { createPost } from "../slices/postsSlice";
import { postPost } from "../services/postPost";
import type { Post } from "@/types/Post";

export function useCreatePost() {
	const [state, setState] = useState<State>("idle");
	const dispatch = useDispatch();

	const create = useCallback(
		async (post: Omit<Post, "id">) => {
			try {
				setState("loading");
				const response = await postPost(post);
				dispatch(createPost(response));
				setState("success");
			} catch (error) {
				setState("error");
			}
		},
		[dispatch],
	);

	return { create, state };
}
