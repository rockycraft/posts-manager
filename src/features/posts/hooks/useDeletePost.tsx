import type { State } from "@/types/State";
import { useState } from "react";
import { deletePost as deletePostService } from "../services/deletePost";
import { useDispatch } from "react-redux";
import { removePost } from "../slices/postsSlice";
import { Post } from "@/types/Post";

export function useDeletePost() {
	const [state, setState] = useState<State>("idle");
	const dispatch = useDispatch();

	async function deletePost(id: Post["id"]) {
		try {
			setState("loading");
			//@ts-ignore
			const response = await deletePostService(id);
			dispatch(removePost(response.id));
			setState("success");
		} catch (error) {
			console.error(error);
			setState("error");
		}
	}

	return { deletePost, state };
}
