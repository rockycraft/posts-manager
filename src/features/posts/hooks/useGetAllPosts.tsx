import type { RootState } from "@/store";
import { useCallback, useEffect, useRef, useState } from "react";
import { shallowEqual, useDispatch, useSelector } from "react-redux";
import { getPosts } from "../services/getPosts";
import { addPosts } from "../slices/postsSlice";
import type { State } from "@/types/State";

export function useGetAllPosts() {
	const [state, setState] = useState<State>("idle");
	const initialized = useRef(false);
	const posts = useSelector(
		(state: RootState) => state.posts.posts,
		shallowEqual,
	);
	const dispatch = useDispatch();

	const getAllPosts = useCallback(async () => {
		try {
			setState("loading");
			const response = await getPosts();
			dispatch(addPosts(response));
			setState("success");
		} catch (error) {
			setState("error");
			console.error(error);
		}
	}, [dispatch]);

	useEffect(() => {
		if (!initialized.current) {
			initialized.current = true;

			getAllPosts();
		}
	}, [getAllPosts]);

	return { data: posts, state, getAllPosts };
}
