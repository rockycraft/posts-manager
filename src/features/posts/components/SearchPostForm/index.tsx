import { type Dispatch, memo, type SetStateAction, useCallback } from "react";
import type { z } from "zod";
import { type SubmitHandler, useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";
import { postSchema } from "@/types/Post";
import type { ColumnFilters } from "../Table";
import Form from "./Form";

const searchPostSchema = postSchema.omit({ id: true, description: true });

type SearchPostInput = z.infer<typeof searchPostSchema>;

type SearchPostFormProps = {
	onSearch: Dispatch<SetStateAction<ColumnFilters>>;
};

function SearchPostForm({ onSearch }: SearchPostFormProps) {
	const { register, handleSubmit } = useForm<SearchPostInput>({
		resolver: zodResolver(searchPostSchema),
	});

	const onSubmit = useCallback<SubmitHandler<SearchPostInput>>(
		(data) => {
			onSearch(() => [
				{
					id: "name",
					value: data.name,
				},
			]);
		},
		[onSearch],
	);

	return <Form onSubmit={handleSubmit(onSubmit)} register={register} />;
}

export default memo(SearchPostForm);
