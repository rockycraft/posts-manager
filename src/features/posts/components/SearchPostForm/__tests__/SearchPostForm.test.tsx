import "@testing-library/jest-dom";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { beforeEach, describe, expect, it, vi } from "vitest";
import SearchPostForm from "../index";
import { useCreatePost } from "../../../hooks/useCreatePost";
import { toaster } from "@/components/ui/toaster";
import { Provider } from "@/components/ui/provider";

Object.defineProperty(window, "matchMedia", {
	writable: true,
	value: vi.fn().mockImplementation((query) => ({
		matches: false,
		media: query,
		onchange: null,
		addListener: vi.fn(), // deprecated
		removeListener: vi.fn(), // deprecated
		addEventListener: vi.fn(),
		removeEventListener: vi.fn(),
		dispatchEvent: vi.fn(),
	})),
});

vi.mock("../../../hooks/useCreatePost");
vi.mock("@/components/ui/toaster", () => ({
	Toaster: () => <div>Toaster</div>,
	toaster: {
		create: vi.fn(),
	},
}));

describe("NewPostForm component", () => {
	const mockCreate = vi.fn();
	const mockReset = vi.fn();
	const mockOnSearch = vi.fn();

	beforeEach(() => {
		vi.clearAllMocks();
		// @ts-ignore
		useCreatePost.mockReturnValue({ create: mockCreate, state: "idle" });
	});

	it("should take a snapshot", () => {
		const { asFragment } = render(
			<Provider>
				<SearchPostForm onSearch={mockOnSearch} />
			</Provider>,
		);
		expect(asFragment()).toMatchSnapshot();
	});

	it("should render the from with inputs and button", () => {
		render(
			<Provider>
				<SearchPostForm onSearch={mockOnSearch} />
			</Provider>,
		);

		expect(screen.getByTestId("name-search-input")).toBeInTheDocument();
		expect(screen.getByTestId("submit-button")).toBeInTheDocument();
	});

	it("should call create and reset on form submit", () => {
		render(
			<Provider>
				<SearchPostForm onSearch={mockOnSearch} />
			</Provider>,
		);

		fireEvent.input(screen.getByTestId("name-search-input"), {
			target: { value: "Test Name" },
		});

		fireEvent.submit(screen.getByRole("form"));

		waitFor(() => {
			expect(mockOnSearch).toHaveBeenCalledTimes(1);
			expect(mockCreate).toHaveBeenCalledTimes(1);
			expect(mockCreate).toHaveBeenCalledWith({
				name: "Test Name",
				description: "Test Description",
			});
			expect(mockReset).toHaveBeenCalledTimes(1);
		});
	});

	it("should show error toaster when state is error", () => {
		//@ts-ignore
		useCreatePost.mockReturnValue({
			create: mockCreate,
			state: "error",
		});

		render(
			<Provider>
				<SearchPostForm onSearch={mockOnSearch} />
			</Provider>,
		);

		waitFor(() => {
			expect(toaster.create).toHaveBeenCalledWith({
				title: "Ocurri√≥ un error al crear el post",
				type: "error",
			});
		});
	});
});
