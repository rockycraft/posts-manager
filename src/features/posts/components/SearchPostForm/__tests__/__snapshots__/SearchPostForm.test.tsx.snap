// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NewPostForm component > should take a snapshot 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <form
    role="form"
  >
    <div
      class="chakra-stack css-1igwmid"
    >
      <input
        class="chakra-input css-1hens3x"
        data-testid="name-search-input"
        name="name"
        placeholder="Nombre"
        type="text"
      />
      <button
        class="chakra-button css-4hnymn"
        data-testid="submit-button"
        type="submit"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M10.5 3.75a6.75 6.75 0 1 0 0 13.5 6.75 6.75 0 0 0 0-13.5ZM2.25 10.5a8.25 8.25 0 1 1 14.59 5.28l4.69 4.69a.75.75 0 1 1-1.06 1.06l-4.69-4.69A8.25 8.25 0 0 1 2.25 10.5Z"
            fill-rule="evenodd"
          />
        </svg>
        Buscar
      </button>
    </div>
  </form>
</DocumentFragment>
`;
